src/mlpack/methods/linear_svm/linear_svm_impl.hpp

```diff
@@ -0,0 +1,165 @@
+/**
+ * @file linear_svm.cpp
+ * @author Ayush Chamoli
+ *
+ * Implementation of Linear SVM.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_LINEAR_SVM_LINEAR_SVM_IMPL_HPP
+#define MLPACK_METHODS_LINEAR_SVM_LINEAR_SVM_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "linear_svm.hpp"
+
+namespace mlpack {
+namespace svm {
+
+template <typename MatType>
+template <typename OptimizerType>
+LinearSVM<MatType>::LinearSVM(
+    const MatType& data,
+    const arma::Row<size_t>& labels,
+    const size_t numClasses,
+    const double lambda,
+    const double delta,
+    OptimizerType optimizer) :
+    numClasses(numClasses),
+    lambda(lambda),
+    delta(delta)
+{
+  Train(data, labels, numClasses, optimizer);
+}
+
+template <typename MatType>
+LinearSVM<MatType>::LinearSVM(
+    const size_t numClasses,
+    const double lambda,
+    const double delta) :
+    numClasses(numClasses),
+    lambda(lambda),
+    delta(delta)
+{
+  // No training to do here.
+}
+
+template <typename MatType>
+template <typename OptimizerType>
+double LinearSVM<MatType>::Train(
+    const MatType& data,
+    const arma::Row<size_t>& labels,
+    const size_t numClasses,
+    OptimizerType optimizer)
+{
+  LinearSVMFunction<MatType> svm(data, labels,
+      numClasses, lambda, delta);
+  if (parameters.is_empty())
+    parameters = svm.InitialPoint();
+
+  // Train the model.
+  Timer::Start("linear_svm_optimization");
+  const double out = optimizer.Optimize(svm, parameters);
+  Timer::Stop("linear_svm_optimization");
+
+  Log::Info << "LinearSVM::LinearSVM(): final objective of "
+            << "trained model is " << out << "." << std::endl;
+
+  return out;
+}
+
+template <typename MatType>
+void LinearSVM<MatType>::Classify(
+    const MatType& data,
+    arma::Row<size_t>& labels)
+const
+{
+  arma::mat scores;
+  Classify(data, labels, scores);
+}
+
+template <typename MatType>
+void LinearSVM<MatType>::Classify(
+    const MatType& data,
+    arma::Row<size_t>& labels,
+    arma::mat& scores)
+const
+{
+  Classify(data, scores);
+
+  // Prepare necessary data
+  labels.zeros(data.n_cols);
+  double maxScore = 0;
+
+  // For each test input.
+  for (size_t i = 0; i < data.n_cols; ++i)
+  {
+    // For each class.
+    for (size_t j = 0; j < numClasses; ++j)
+    {
+      // If a higher class probability is encountered, change score.
+      if (scores(j, i) > maxScore)
+      {
+        maxScore = scores(j, i);
+        labels(i) = j;
+      }
+    }
+
+    // Set maximum probability to zero for next input.
+    maxScore = 0;
+  }
```

Comment:
Could `arma::index_max()` be used here?  Note that that is only available on Armadillo 7.300.0 and newer, but you can use things like `#if ARMA_VERSION_MAJOR > 7 || ARMA_VERSION_MAJOR == 7 && ARMA_VERSION_MINOR >= 300` to choose which to use.  Eventually as the minimum version of Armadillo increases we can remove the code that doesn't use `index_max()`.

Reply:
I didn't go through this but I tried the program that was available in the documentation of `arma::index_max()`. And it basically gave the linear index of the maximum element inside the matrix (could have realized that if I had just read the documentation, but I kinda skipped that, and ran the example code :rofl:). 
I have not given this a complete thought but I think we will have to use `arma::index_max()` on subsets of scores to find the max index of each column of the `dataset`. I think we can have this code for now, but it can definately be implemented in the way you suggested (Also, since `SoftmaxRegression` uses similar code, we can use `arma::index_max()` over there too).

Reply:
I'm pretty sure the standalone `index_max()` will, on matrices, give the index of the maximum element for each row or column: http://arma.sourceforge.net/docs.html#index_min_and_index_max_standalone

(Note that's different than the member `.index_max()` function.)

Reply:
Oh, ok, sorry for my misinterpretation. I think I read about `.index_max()` member function.